runtime/sema/check_transaction_declaration.go

```diff
@@ -248,36 +239,168 @@ func (checker *Checker) declareTransactionDeclaration(declaration *ast.Transacti
 		transactionType.Parameters = checker.parameters(declaration.ParameterList)
 	}
 
+	members, fields, origins := checker.compositeFieldMembersAndOrigins(
+		transactionType,
+		declaration.DeclarationKind(),
+		declaration.Fields,
+	)
+
+	transactionType.Members = members
+	transactionType.Fields = fields
+	if checker.PositionInfo != nil {
+		checker.PositionInfo.recordMemberOrigins(transactionType, origins)
+	}
+
+	if declaration.Prepare != nil {
+		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
+		transactionType.PrepareParameters = checker.parameters(parameterList)
+	}
+
+	roles := &orderedmap.OrderedMap[string, *TransactionRoleType]{}
+	for _, roleDeclaration := range declaration.Roles {
+		transactionRoleType := checker.transactionRoleType(roleDeclaration)
+		checker.Elaboration.SetTransactionRoleDeclarationType(roleDeclaration, transactionRoleType)
```

Comment:
Should we move this after the duplicate/field name clash checking? It seems a waste to compute the type if we are just going to discard it when we `continue`.

Reply:
We always want to report as much as possible. Just because the field is duplicated or there is a clash with a field should not prevent us from also checking the transaction role itself later on.

Reply:
Agreed, but unless I am mistaken because we have `continue` statements in the conditionals that check for name conflicts, we are not actually adding the newly computed `role` type to the transaction's `role` fields at the end of the loop, so we aren't going to typecheck it regardless.

Reply:
Even if roles are duplicated or clash with fields, still all AST role declarations are iterated and checked (see https://github.com/onflow/cadence/pull/2257/files/a5ca45aba8ec70e8cb41a4bfe7e9dcbd36c6e581#diff-3ee2ecce28d4ab69f08348e38d2c38de4610023f91f00b769787df1d5a14e07aR73-R75 and in turn `VisitTransactionRoleDeclaration`), which assumes that the transaction role type was defined.

You can check this for yourself by attempting the proposed change, moving the transaction type creation and the association of the AST role declaration with the role type in the elaboration below the continue statements of the duplication/clash checks.

Reply:
I still improved it a bit in 149085186089642b1f395c39ca7465a422be4c7f by replacing the continue statements with separate checks for duplication and field clashing, and in each case, performing the operation (associating the role if not a duplicate, and declaring a member if no field clash) separately if possible.